### PR TITLE
Style and lint

### PR DESCRIPTION
**Pull request by philoserf**
This PR was constructed as I reviewed the code and built my understanding of the codebase. You can think of it as the work of an editor.

- Apply Dockerfile best practices:https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#add-or-copy
- Use reference style links. I find them easier to type, to read in plain text, and to maintain. For example in the original there were two links to the kraken repo. Now there are two references to one link. https://daringfireball.net/projects/markdown/syntax#link
- standardize shell script string quoting. Quoting the full variable containing string is cleaner than quoting individual variables. I find it looks more like the standard string interpolation we are used to win other languages.
- Use shellcheck to help me find potential issue and to leverage the wisdom of a larger community. see: https://github.com/koalaman/shellcheck and http://www.shellcheck.net
- Use more of the Google Shell Style Guide according to stated team standard.
- See more editorial comments annotated with the code.

---
This change is [Reviewable](https://reviewable.io/reviews/samsung-cnct/kraken-tools/145)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -44,7 +44,7 @@ ENV     APP_REGISTRY_URL=https://github.com/app-registry/appr-helm-plugin/releas
 
 # Alpine
 
-ADD     build/alpine-builds /alpine-builds
+COPY     build/alpine-builds /alpine-builds
```
Comment (philoserf) [inline Dockerfile]: https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#add-or-copy

File: Jenkinsfile [hunk 2]
```diff
@@ -13,46 +13,45 @@ podTemplate(label: 'kraken-tools', containers: [
...
-            stage('Publish') {
-              if (git_branch.contains(publish_branch) && git_uri.contains(github_org)) {
-                kubesh "docker tag kraken-tools:${env.JOB_BASE_NAME}.${env.BUILD_ID} quay.io/${quay_org}/k2-tools:${image_tag}"
-                kubesh "docker tag kraken-tools:${env.JOB_BASE_NAME}.${env.BUILD_ID} quay.io/${quay_org}/kraken-tools:${image_tag}"
-                kubesh "docker push quay.io/${quay_org}/k2-tools:${image_tag}"
-                kubesh "docker push quay.io/${quay_org}/kraken-tools:${image_tag}"
-              } else {
-                echo "Not pushing to docker repo:\n    BRANCH_NAME='${env.BRANCH_NAME}'\n    GIT_BRANCH='${git_branch}'\n    git_uri='${git_uri}'"
-              }
+      customContainer('docker') {
```
Comment (philoserf) [inline Jenkinsfile]: Use consistent indentation. Above and below this section, we use 2 spaces. Here we used 4 spaces.

**Comment by leahnp** [comment]
@philoserf I am working on: https://github.com/samsung-cnct/issues/issues/68 currently and will be producing a doc and several examples of your first bullet point: `Apply Dockerfile best practices`. My PR's on this point should start later today, I'd love your feedback on them.

File: README.md [hunk 3]
```diff
@@ -1,30 +1,57 @@
 # kraken-tools
 
-kraken-tools conveniently manages all system requirements for [kraken-lib](https://github.com/samsung-cnct/kraken-lib). As the base layer, it contains all necessary and expected dependencies at specific, tested versions for correct and repeatable operations of kraken-lib and [kraken](https://github.com/samsung-cnct/kraken). 
+kraken-tools conveniently manages all system requirements for [kraken-lib][kl].
```
Comment (philoserf) [inline README.md]: 80 character lines work well for command line users and flow just fine when rendered by _nearly_ all markdown rendering engines.

File: build/alpine-builds/ansible/APKBUILD [hunk 4]
```diff
@@ -13,14 +14,14 @@ makedepends="python-dev py-setuptools"
...
+builddir="$srcdir/$pkgname-$pkgver"
 build() {
-	cd "$builddir"
+	cd "$builddir" || exit 1
 	python setup.py build || return 1
 }
 
 package() {
-	cd "$builddir"
+	cd "$builddir" || exit 1
```
Comment (philoserf) [inline build/alpine-builds/ansible/APKBUILD]: the rationale for ``|| ext` is here https://github.com/koalaman/shellcheck/wiki/SC2164

File: build/alpine-builds/ansible/APKBUILD [hunk 5]
```diff
@@ -13,14 +14,14 @@ makedepends="python-dev py-setuptools"
 subpackages="$pkgname-doc"
 source="$pkgname-$pkgver-0.2.rc2.tar.gz::http://releases.ansible.com/ansible/$pkgname-$pkgver-0.2.rc2.tar.gz"
 
-builddir="$srcdir"/$pkgname-$pkgver
+builddir="$srcdir/$pkgname-$pkgver"
```
Comment (philoserf) [inline build/alpine-builds/ansible/APKBUILD]: Quoting the full variable containing string is cleaner than quoting individual variables.
I find it looks more like the standard string interpolation we are used to win other languages.

File: build/alpine-builds/ansible/APKBUILD [hunk 6]
```diff
@@ -34,9 +35,16 @@ package() {
...
+		"$pkgdir/usr/share/doc/$pkgname/README.core.md" || return 1
 	install -m644 lib/ansible/modules/extras/README.md \
-		"$pkgdir"/usr/share/doc/$pkgname/README.extras.md || return 1
+		"$pkgdir/usr/share/doc/$pkgname/README.extras.md" || return 1
+}
+
+# WARNING: ansible*: APKBUILD does not run any tests!
+#    Alpine policy will soon require that packages have any relevant testsuites run during the build process.
+#    To fix, either define a check() function, or declare !check in $options to indicate the package does not have a testsuite.
+check() { # placeholder
```
Comment (philoserf) [inline build/alpine-builds/ansible/APKBUILD]: This surpasses the warning in the build and serves as a marker to future readers that something could be added to improve the APKBUILD.

File: build/alpine-builds/build-docker.sh [hunk 7]
```diff
@@ -1,3 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
Comment (philoserf) [inline build/alpine-builds/build-docker.sh]: The Google Shell Style Guide is prescriptive about `#!/bin/bash`.

File: build/alpine-builds/build.sh [hunk 8]
```diff
@@ -1,8 +1,9 @@
-#!/bin/sh
+#!/bin/bash
+
 apk update
 apk add alpine-sdk
 adduser -D samsung
-echo "\nsamsung ALL=(ALL) ALL\n" > /etc/sudoers
+printf "\nsamsung ALL=(ALL) ALL\n" > /etc/sudoers
```
Comment (philoserf) [inline build/alpine-builds/build.sh]: `printf` over `echo` for new line handling

File: build/gcloud_tree.py [hunk 9]
```diff
@@ -1,10 +1,5 @@
 #!/bin/env python
 
-# The original gcloud_tree is generated with the following command:
```
Comment (philoserf) [inline build/gcloud_tree.py]: The comment deserves to be in the pull request that proposed the change. Today I tells the reader about uninteresting history.

File: tests/aws-testing.sh [hunk 10]
```diff
@@ -1,40 +1,42 @@
-#!/bin/sh
+#!/bin/bash
+
 apk update
 apk add git
-git clone --branch master  https://github.com/samsung-cnct/kraken-lib.git ~/kraken
+git clone --branch master  https://github.com/samsung-cnct/kraken-lib.git "$HOME/kraken"
```
Comment (philoserf) [inline tests/aws-testing.sh]: `$HOME` is safer than `~` for home folder expansion in a script. for example were some future change require the line be quoted then the `~` would not be expanded.

File: tests/aws-testing.sh [hunk 11]
```diff
@@ -1,40 +1,42 @@
...
 echo "prep complete, create config"
 bin/up.sh --generate cluster/aws/config.yaml
 
 echo "modify config in-place"
-build-scripts/update-generated-config.sh cluster/aws/config.yaml ${JOB_BASE_NAME}-${BUILD_ID}
+build-scripts/update-generated-config.sh cluster/aws/config.yaml "${JOB_BASE_NAME}-${BUILD_ID}"
 
 echo "adding helm_override and setting it to false"
-job_name=`echo ${JOB_BASE_NAME}-${BUILD_ID} | tr '[:upper:]' '[:lower:]' | tr '-' '_'`
+job_name=$(echo "${JOB_BASE_NAME}-${BUILD_ID}" | tr '[:upper:]' '[:lower:]' | tr '-' '_')
```
Comment (philoserf) [inline tests/aws-testing.sh]: shellcheck justification for $(...) over `...` https://github.com/koalaman/shellcheck/wiki/SC2006

File: tests/aws-testing.sh [hunk 12]
```diff
@@ -1,40 +1,42 @@
...
 export helm_override_${job_name}=false
 
 echo "dry run"
-PWD=`pwd` && bin/up.sh --config $PWD/cluster/aws/config.yaml --output $PWD/cluster/aws/ -t dryrun
-if [ $? -ne 0 ]; then
+PWD=$(pwd) && bin/up.sh --config "$PWD/cluster/aws/config.yaml" --output "$PWD/cluster/aws/" -t dryrun
```
Comment (philoserf) [inline tests/aws-testing.sh]: I have to look at this closer. Seems weird. May have something in a different PR. until then follow the `$()` model.

File: tests/aws-testing.sh [hunk 13]
```diff
@@ -1,40 +1,42 @@
...
+dryrun_err=$?
```
Comment (philoserf) [inline tests/aws-testing.sh]: For consistency with `down_err` and `up_err`. Also shellcheck justification https://github.com/koalaman/shellcheck/wiki/SC2202

File: tests/internal_tooling.sh [hunk 14]
```diff
@@ -4,24 +4,24 @@
 # NOTE this script leverages a few bash-isms that may not work in other shells.
 
 matches_version () {
-    egrep -q '^v[0-9]+\.[0-9]+$'
+    grep -q -E '^v[0-9]+\.[0-9]+$'
```
Comment (philoserf) [inline tests/internal_tooling.sh]: shellcheck justification for `grep -E` over `egrep` https://github.com/koalaman/shellcheck/wiki/SC2196

File: tests/internal_tooling.sh [hunk 15]
```diff
@@ -4,24 +4,24 @@
...
     find /opt/cnct/kubernetes -maxdepth 1 -type d -name 'v*.*' | \
-        egrep 'v[0-9]+\.[0-9]+$'
+        grep -E 'v[0-9]+\.[0-9]+$'
 }
 
 
 kube_path_contains_binaries () {
     count=0
-    while read versionpath; do
+    while read -r versionpath; do
```
Comment (philoserf) [inline tests/internal_tooling.sh]: shellcheck justification for `read -r` over bare `read` https://github.com/koalaman/shellcheck/wiki/SC2162

**Comment by philoserf** [comment]
Looks like Jenkins found issues my local build couldn't. I'll dig in.

In the meantime II believe this PR is could be a good conversation about editing code. I'll have a blog post up later this week comparing code to essays, articles, short stories, and other kinds of writing.

TLDR of the as yet unwritten blog post:

Code is an essay written over time. Occasionally, even commonly, written with laser focus on one paragraph at a time, or one word at a time even.
